community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala

```diff
@@ -22,34 +22,82 @@ import scala.collection.mutable.ArrayBuffer
 
 trait Attribute[T] {
 
-  // TODO crappy perf?
-  val array:ArrayBuffer[Unchangeable[T]] = new ArrayBuffer[Unchangeable[T]]()
+  private val array: ArrayBuffer[Unchangeable[T]] = new ArrayBuffer[Unchangeable[T]]()
 
-  def set(id:Id, t:T): Unit = {
+  def set(id: Id, t: T): Unit = {
     val requiredSize = id.x + 1
     if (array.size < requiredSize)
       resizeArray(requiredSize)
     array(id.x).value = t
   }
 
-  def get(id:Id): T = {
+  def get(id: Id): T = {
     array(id.x).value
   }
 
-  def copy(from:Id, to:Id): Unit = {
+  def isDefinedAt(id: Id): Boolean = {
+    array.size > id.x && array(id.x).hasValue
+  }
+
+  def getOrElse(id: Id, other: => T): T = {
+    if (isDefinedAt(id)) get(id) else other
+  }
+
+  def iterator: Iterator[(Id, T)] = new Iterator[(Id, T)]() {
+    private var currentId = -1
+    private var nextTup: (Id, T) = _
+
+    private def fetchNext(): Unit = {
+      nextTup = null
+      while (nextTup == null &&
+        {currentId = currentId + 1; currentId} < array.size) {
+        val c = array(currentId)
+        if (c.hasValue) {
+          nextTup = (Id(currentId), c.value)
+        }
+      }
+    }
+
+    override def hasNext = {
+      if (currentId >= array.size)
+        false
+      else {
+        if (nextTup == null) {
+          fetchNext()
+        }
+        nextTup != null
+      }
+    }
+
+    override def next() = {
+      if (hasNext) {
+        val res = nextTup
+        nextTup = null
+        res
+      } else {
+        throw new NoSuchElementException
+      }
+    }
+  }
+
+  def size = iterator.size
```

Comment:
If we ever think this is slow, we could just keep a private size field which is incremented on `set` if there was no previous value.